File: 프로그래머스/2/1844. 게임 맵 최단거리/게임 맵 최단거리.js
function solution(maps) {
    var answer = 0;
    const n = maps.length;
    const m = maps[0].length;
    
    // 최단 거리 문제, BFS
    let ch = Array.from(Array(n) , ()=>Array(m).fill(0)) // 체크 배열
    let queue=[];
    
    // 동서남북 방향
    const dx = [0,0,1,-1];
    const dy = [1,-1,0,0];
  
    queue.push([0,0,1]);
    ch[0][0] = 1;
    
    while (queue.length > 0) {
        
        const [sx,sy,ans] = queue.shift();
        
        if (sx == n-1 && sy == m-1) {
            // 도착 가능
            return ans;
        } 
        
        for (let i=0 ; i < 4; i++) {
            let x = dx[i] + sx;
            let y = dy[i] + sy;
            
            if (x >= 0 && y >= 0 && x < n && y < m && maps[x][y] == 1 && ch[x][y] == 0){
                // 이동 가능
                ch[x][y] = 1;
                queue.push([x,y,ans+1]);
            }
        }
        
    }
     return -1;
    
}

File: 프로그래머스/2/1844. 게임 맵 최단거리/게임 맵 최단거리.py
from collections import deque

def solution(maps):
    limit = int(1e9)
    
    answer = limit
    n = len(maps)
    m = len(maps[0])
    
    visit = [[False] * (m+1) for _ in range(n+1)]
    queue = deque([(0,0,1)])
    visit[0][0] = True
    
    dx=[-1,1,0,0]
    dy=[0,0,-1,1]
    
    while(queue):
        x, y, cnt = queue.popleft()
        
        if x == n-1 and y == m-1:
            # 도착
            answer = min(answer, cnt)
        
        for i in range(4):
            nx = x + dx[i]
            ny = y + dy[i]
            
            if 0 <= nx < n and 0 <= ny < m and visit[nx][ny] == False and maps[nx][ny] == 1:
                visit[nx][ny] = True
                queue.append((nx,ny, cnt+1))
    
    if answer == limit :
        return -1 
    else:
        return answer

